libgalois/src/DynamicBitset.cpp

```diff
@@ -41,6 +41,12 @@ katana::DynamicBitset::bitwise_not() {
   katana::do_all(
       katana::iterate(size_t{0}, bitvec_.size()),
       [&](size_t i) { bitvec_[i] = ~bitvec_[i]; }, katana::no_stats());
+
+  // invariant: the unused bits in the last element of bitvec_ are always 0
+  size_t last_entry_offset =
```

Comment:
Would it make sense to pull this out in a function like "mask non zeros" (or whatever) and then it is a little more obvious why only "not" needs to do this?

Reply:
Why would pulling this out into a function make it more clear why only "not" needs to do it?

Reply:
why don't we need to mask the result of xor?

Reply:
We only support `xor` on bitsets of the same size. And `xor` will maintain the invariant. If you pass in a bitset where all unused bits are 0, you will get out a bitset where all unused bits are 0.

Reply:
> Why would pulling this out into a function make it more clear why only "not" needs to do it?

Because you can write this:

```cpp
/// MaskTrailingNonZeros sets any non-zeros outside of the nominal size of the bitset to zero. Count
/// depends on such extra bits to be zero to return the correct value.
//
// As a micro-optimization, we can skip calling this when the trailing bits are known to be zero by construction.
void MaskTrailingNonZeros() {
 ...
}

void bitwise_not(...) {
  ...
  MaskTrailingNonZeros();
}
```

Though, I'd just call `MaskTrailingNonZeros` all the time because it is basically free to call and ensures that we maintain the class invariant now and in the future.